### ISSUE TITLE
On upgrade to 20151015T063628Z KVM VM failed to boot (NIC definition was missing the "ips" property)

### ISSUE THREAD

**Issue by nshalman** [joyent/smartos-live#493]
(upgrading from 20150905T134251Z)

I had set autoboot=false on all of my KVM VMs. All my other zones appear to have booted just fine (and been updated?)

I tried to boot my main KVM VM and vmadm dumped core after squeaking out:
```
Uncaught Error: expected an array of IP addresses: undefined
```

As a workaround I manually edited the /etc/zones/uuid.xml for the VM and just below:
```
[snip]
  <network physical="net0" mac-addr="[snip]" global-nic="admin">
    <net-attr name="gateway" value="dhcp"/>
    <net-attr name="ip" value="dhcp"/>
```
I added:
```
    <net-attr name="ips" value="dhcp"/>
```
Which allowed that VM to boot.

**Comment by rmustacc** [comment]
@nshalman Is it possible to provide what the VM used to look like by chance?

@melloc Can you take a look at this?

**Comment by nshalman** [comment]
@rmustacc I'm not clear on what you want. Do you want the XML file content or the ```vmadm get``` output or what? I suspect @melloc will be able to reproduce it pretty easily, but am happy to provide whatever additional information I can that will be helpful.

**Comment by nshalman** [comment]
I had a second VM that was failing in the same way. Adding in the extra line (the "ips" line) fixed it, and removing the extra line broke it again.

So perhaps two issues:
1. Some zone config files aren't getting upgraded.
2. Un-upgraded zone config files make vmadm dump core.

**Comment by rmustacc** [comment]
@nshalman In this case can you provide the core file and the xml output?

**Comment by melloc** [comment]
@nshalman It looks like your `gateway` value is set to `dhcp`, which is probably preventing the upgrade. If you remove that line and restart `vmadmd`, it should upgrade correctly. I'm guessing it's dumping core because some path is expecting the upgrade to have happened successfully. Can you share the core file? I'll harden vmadm to handle the upgrade failure better.

**Comment by melloc** [comment]
@nshalman You might also be able to take a look at vmadmd's logs (`svcs -L vmadmd`) to get a better idea of why they're not getting upgraded.

**Comment by sjorge** [comment]
FWIW I didn't have this issue with my 2 kvm zones, but they both run with a static IP/no gateway set.

**Comment by nshalman** [comment]
I have a core file if it's still desired. Find me on IRC and I'll send it along.

Looks like it's sad about ```"netmask":"dhcp"```:
```
# grep -i upgrad `svcs -L vmadmd`
{"name":"vmadmd","hostname":"smartos","pid":16182,"level":50,"err":{"message":"Specified netmask value is invalid: dhcp","name":"Error","stack":"Error: Specified netmask value is invalid: dhcp\n    at upgradeNic (/usr/vm/node_modules/nic.js:39:18)\n    at /usr/node/0.10/node_modules/async.js:235:13\n    at /usr/node/0.10/node_modules/async.js:118:13\n    at Array.forEach (native)\n    at _each (/usr/node/0.10/node_modules/async.js:39:24)\n    at async.each (/usr/node/0.10/node_modules/async.js:117:9)\n    at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)\n    at Object.map (/usr/node/0.10/node_modules/async.js:212:23)\n    at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)\n    at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)"},"msg":"Specified netmask value is invalid: dhcp","time":"2015-10-16T13:37:47.458Z","v":0}
{"name":"vmadmd","hostname":"smartos","pid":16182,"level":50,"err":{"message":"Specified netmask value is invalid: dhcp","name":"Error","stack":"Error: Specified netmask value is invalid: dhcp\n    at upgradeNic (/usr/vm/node_modules/nic.js:39:18)\n    at /usr/node/0.10/node_modules/async.js:235:13\n    at /usr/node/0.10/node_modules/async.js:118:13\n    at Array.forEach (native)\n    at _each (/usr/node/0.10/node_modules/async.js:39:24)\n    at async.each (/usr/node/0.10/node_modules/async.js:117:9)\n    at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)\n    at Object.map (/usr/node/0.10/node_modules/async.js:212:23)\n    at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)\n    at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)"},"msg":"Specified netmask value is invalid: dhcp","time":"2015-10-16T13:37:48.114Z","v":0}
{"name":"vmadmd","hostname":"smartos","pid":16182,"level":50,"err":{"message":"Specified netmask value is invalid: dhcp","name":"Error","stack":"Error: Specified netmask value is invalid: dhcp\n    at upgradeNic (/usr/vm/node_modules/nic.js:39:18)\n    at /usr/node/0.10/node_modules/async.js:235:13\n    at /usr/node/0.10/node_modules/async.js:118:13\n    at Array.forEach (native)\n    at _each (/usr/node/0.10/node_modules/async.js:39:24)\n    at async.each (/usr/node/0.10/node_modules/async.js:117:9)\n    at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)\n    at Object.map (/usr/node/0.10/node_modules/async.js:212:23)\n    at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)\n    at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)"},"msg":"Specified netmask value is invalid: dhcp","time":"2015-10-16T13:37:48.741Z","v":0}
{"name":"vmadmd","hostname":"smartos","pid":16182,"level":50,"err":{"message":"Specified netmask value is invalid: dhcp","name":"Error","stack":"Error: Specified netmask value is invalid: dhcp\n    at upgradeNic (/usr/vm/node_modules/nic.js:39:18)\n    at /usr/node/0.10/node_modules/async.js:235:13\n    at /usr/node/0.10/node_modules/async.js:118:13\n    at Array.forEach (native)\n    at _each (/usr/node/0.10/node_modules/async.js:39:24)\n    at async.each (/usr/node/0.10/node_modules/async.js:117:9)\n    at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)\n    at Object.map (/usr/node/0.10/node_modules/async.js:212:23)\n    at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)\n    at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)"},"msg":"Specified netmask value is invalid: dhcp","time":"2015-10-16T13:37:49.388Z","v":0}
{"name":"vmadmd","hostname":"smartos","pid":16235,"level":50,"err":{"message":"Specified netmask value is invalid: dhcp","name":"Error","stack":"Error: Specified netmask value is invalid: dhcp\n    at upgradeNic (/usr/vm/node_modules/nic.js:39:18)\n    at /usr/node/0.10/node_modules/async.js:235:13\n    at /usr/node/0.10/node_modules/async.js:118:13\n    at Array.forEach (native)\n    at _each (/usr/node/0.10/node_modules/async.js:39:24)\n    at async.each (/usr/node/0.10/node_modules/async.js:117:9)\n    at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)\n    at Object.map (/usr/node/0.10/node_modules/async.js:212:23)\n    at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)\n    at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)"},"msg":"Specified netmask value is invalid: dhcp","time":"2015-10-16T13:38:35.794Z","v":0}
{"name":"vmadmd","hostname":"smartos","pid":16235,"level":50,"err":{"message":"Specified netmask value is invalid: dhcp","name":"Error","stack":"Error: Specified netmask value is invalid: dhcp\n    at upgradeNic (/usr/vm/node_modules/nic.js:39:18)\n    at /usr/node/0.10/node_modules/async.js:235:13\n    at /usr/node/0.10/node_modules/async.js:118:13\n    at Array.forEach (native)\n    at _each (/usr/node/0.10/node_modules/async.js:39:24)\n    at async.each (/usr/node/0.10/node_modules/async.js:117:9)\n    at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)\n    at Object.map (/usr/node/0.10/node_modules/async.js:212:23)\n    at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)\n    at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)"},"msg":"Specified netmask value is invalid: dhcp","time":"2015-10-16T13:38:36.445Z","v":0}
{"name":"vmadmd","hostname":"smartos","pid":16235,"level":50,"err":{"message":"Specified netmask value is invalid: dhcp","name":"Error","stack":"Error: Specified netmask value is invalid: dhcp\n    at upgradeNic (/usr/vm/node_modules/nic.js:39:18)\n    at /usr/node/0.10/node_modules/async.js:235:13\n    at /usr/node/0.10/node_modules/async.js:118:13\n    at Array.forEach (native)\n    at _each (/usr/node/0.10/node_modules/async.js:39:24)\n    at async.each (/usr/node/0.10/node_modules/async.js:117:9)\n    at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)\n    at Object.map (/usr/node/0.10/node_modules/async.js:212:23)\n    at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)\n    at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)"},"msg":"Specified netmask value is invalid: dhcp","time":"2015-10-16T13:38:37.077Z","v":0}
{"name":"vmadmd","hostname":"smartos","pid":16235,"level":50,"err":{"message":"Specified netmask value is invalid: dhcp","name":"Error","stack":"Error: Specified netmask value is invalid: dhcp\n    at upgradeNic (/usr/vm/node_modules/nic.js:39:18)\n    at /usr/node/0.10/node_modules/async.js:235:13\n    at /usr/node/0.10/node_modules/async.js:118:13\n    at Array.forEach (native)\n    at _each (/usr/node/0.10/node_modules/async.js:39:24)\n    at async.each (/usr/node/0.10/node_modules/async.js:117:9)\n    at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)\n    at Object.map (/usr/node/0.10/node_modules/async.js:212:23)\n    at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)\n    at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)"},"msg":"Specified netmask value is invalid: dhcp","time":"2015-10-16T13:38:37.724Z","v":0}
```

Here's the XML of a "broken" VM (MAC address altered because I'm paranoid like that):
```
<?xml version="1.0" encoding="UTF-8"?>
<!--
    DO NOT EDIT THIS FILE.  Use zonecfg(1M) instead.
-->
<!DOCTYPE zone PUBLIC "-//Sun Microsystems Inc//DTD Zones//EN" "file:///usr/share/lib/xml/dtd/zonecfg.dtd.1">
<zone name="860166be-b54b-48d0-828e-4b6245a42ab0" zonepath="/zones/860166be-b54b-48d0-828e-4b6245a42ab0" autoboot="false" brand="kvm" ip-type="exclusive" limitpriv="default,-file_link_any,-net_access,-proc_fork,-proc_info,-proc_session" debugid="178">
  <attr name="vm-version" type="string" value="1"/>
  <attr name="create-timestamp" type="string" value="2015-01-11T20:14:10.652Z"/>
  <rctl name="zone.cpu-shares">
    <rctl-value priv="privileged" limit="100" action="none"/>
  </rctl>
  <rctl name="zone.zfs-io-priority">
    <rctl-value priv="privileged" limit="100" action="none"/>
  </rctl>
  <rctl name="zone.max-lwps">
    <rctl-value priv="privileged" limit="2000" action="deny"/>
  </rctl>
  <rctl name="zone.max-physical-memory">
    <rctl-value priv="privileged" limit="3221225472" action="deny"/>
  </rctl>
  <rctl name="zone.max-locked-memory">
    <rctl-value priv="privileged" limit="3221225472" action="deny"/>
  </rctl>
  <rctl name="zone.max-swap">
    <rctl-value priv="privileged" limit="3221225472" action="deny"/>
  </rctl>
  <attr name="billing-id" type="string" value="00000000-0000-0000-0000-000000000000"/>
  <attr name="owner-uuid" type="string" value="00000000-0000-0000-0000-000000000000"/>
  <attr name="alias" type="string" value="c21hcnRvcy1seC10ZXN0"/>
  <attr name="ram" type="string" value="2048"/>
  <device match="/dev/zvol/rdsk/zones/860166be-b54b-48d0-828e-4b6245a42ab0-disk0">
    <net-attr name="boot" value="true"/>
    <net-attr name="model" value="virtio"/>
    <net-attr name="media" value="disk"/>
    <net-attr name="image-size" value="10240"/>
    <net-attr name="image-uuid" value="3d09ec02-99c8-11e4-b1a2-8b0aedd63024"/>
  </device>
  <network physical="net0" mac-addr="12:34:56:fe:dc:ba" global-nic="admin">
    <net-attr name="gateway" value="dhcp"/>
    <net-attr name="ip" value="dhcp"/>
    <net-attr name="netmask" value="dhcp"/>
    <net-attr name="model" value="virtio"/>
    <net-attr name="primary" value="true"/>
  </network>
  <attr name="vcpus" type="string" value="3"/>
  <attr name="vm-autoboot" type="string" value="false"/>
</zone>
```

From some manual testing it looks like removing *both* the ```gateway``` and ```netmask``` lines is required to allow the configuration to be upgraded correctly.

And further log inspection (thank you bunyan) yields:
```
[2015-10-16T13:45:09.856Z] ERROR: vmadmd/16288 on smartos: Specified netmask value is invalid: dhcp
    Error: Specified netmask value is invalid: dhcp
        at upgradeNic (/usr/vm/node_modules/nic.js:39:18)
        at /usr/node/0.10/node_modules/async.js:235:13
        at /usr/node/0.10/node_modules/async.js:118:13
        at Array.forEach (native)
        at _each (/usr/node/0.10/node_modules/async.js:39:24)
        at async.each (/usr/node/0.10/node_modules/async.js:117:9)
        at _asyncMap (/usr/node/0.10/node_modules/async.js:234:9)
        at Object.map (/usr/node/0.10/node_modules/async.js:212:23)
        at upgradeNicProp (/usr/vm/node_modules/nic.js:158:15)
        at Object.upgradeNics (/usr/vm/node_modules/nic.js:177:5)
[2015-10-16T13:45:10.788Z] ERROR: vmadmd/16288 on smartos: 
    Error: Static gateways must be IPv4 addresses: dhcp
        at checkPayloadProperties (/usr/vm/node_modules/VM.js:4455:42)
        at Object.vasync.pipeline.funcs [as func] (/usr/vm/node_modules/VM.js:8252:21)
        at Object._onImmediate (/usr/vm/node_modules/vasync/lib/vasync.js:213:20)
        at processImmediate [as _immediateCallback] (timers.js:330:15)
    --
    payload: {
      "update_nics": [
        {
          "mac": "12:34:56:fe:dc:ba",
          "ips": [
            "dhcp"
          ],
          "gateways": [
            "dhcp"
          ]
        }
      ]
    }
```

Let me know if you still need the core file.

**Comment by melloc** [comment]
Resolved by 42a6ad0. Upgrading to the next release shouldn't have this issue.
